### DISCUSSION TITLE
NearCache that only resets AFTER re-established connection

### DISCUSSION THREAD

**Discussion by leolegenie** [oracle/coherence#99] [Q&A]
Hello

Currently the NearCache implementation, if it looses the connection to the backing cluster, does a call to "resetFrontMap()" and drops all data that it contains and all subsequent calls on it then trigger remote calls that hang until the network timeout is reached.

For our usecase, what we would prefer as behavior is that if the NearCache looses the connection to the cluster it goes into a kind of "stale-mode" and keeps at least the data that it already has and returns that to the consumer, as well as returning null for records it does not have.

The goal here is to increase resillience and robustness for the consumer application in case of (temporary) network outages, which we unfortunately have from time to time.

I understand that the LISTEN_NONE strategy does this more or less, but then we have no data updates at all pushed to the near cache from the cluster. So what we would like to have is a LISTEN_ALL strategy but a bit more relaxed on consistency guarantees during network outages.

I tried to implement this behavior by extending NearCache and overriding the DeactivationListener and BackServiceListener and a couple of methods here and there and it seems to work, but I'm not super confident that I got it all correct. I also had to introduce a "connection state" variable to avoid the remote calls when we have no data to return on a get(...) when the connection is down. This in turn seems to impact the BackServiceListener as no MemberJoined event is fired anymore as no reconnect is triggered so I also had to add a poller Thread to re-trigger the connection in the background. So all in all it feels a bit convoluted/fragile. This is on Coherence 14 by the way.

Questions:

1) Is this the right approach to implement such a behavior or is there a better/easier way?
2) How can this be tested properly (currently I do it by running test code manually and influencing the network quality with a tool called "clumsy" - http://jagt.github.io/clumsy/)?
3) Would this be something considered as a future new feature?

Thanks, regards, Leo

**Comment by aseovic** [comment]
Considering that `CQC` already supports disconnected clients, it probably makes sense to provide (optional) support for this in the `NearCache` as well.

If you would like to submit a PR with the work you've done, we'll be happy to work with you on getting it cleaned up, polished and integrated. You can also take a look at how `CQC` handles disconnect and reconnect -- the more both of these are aligned, the better, as they serve similar purpose.

**Reply by leolegenie** [reply]
Ok that sounds good, let me have a look at `CQC`.

Regarding PR: I assume if I'd like to have it integrated into Coherence 14 I should branch out from v14.1.1.0 and then merge against p4-integ-v14.1.10 or how does that work with the branching?
